File: desktopclient/src/main/java/com/goexp/galgame/gui/util/TabSelect.java
package com.goexp.galgame.gui.util;

import javafx.scene.control.TabPane;

import java.util.function.Consumer;

public class TabSelect {

    private Consumer notFind;

    private TabPane root;

    private TabSelect(TabPane root) {
        this.root = root;
    }

    public static TabSelect from(TabPane root) {
        return new TabSelect(root);
    }

    public TabSelect ifNotFind(Consumer notFind) {
        this.notFind = notFind;

        return this;
    }

    public void select(String text) {
        var target = root
                .getTabs().stream()
                .filter(targetTab -> targetTab.getText().equals(text))
                .findAny();

        if (target.isPresent()) {
            root.getSelectionModel().select(target.get());
        } else {
            if (notFind != null)
                notFind.accept(text);
        }
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/util/TabSelect.scala
package com.goexp.galgame.gui.util

import com.goexp.galgame.gui.view.MainController
import javafx.scene.control.{Tab, TabPane}

import scala.jdk.CollectionConverters._

object TabSelect {
  def apply(root: TabPane): TabSelect = new TabSelect(root)

  def apply(): TabSelect = new TabSelect(MainController().homeController.mainTabPanel)
}

class TabSelect private(val root: TabPane) {

  type NotFindAction = () => Tab
  type DataLoader = () => Unit

  private var notFoundAction: NotFindAction = _
  private var loadData: DataLoader = _

  def whenNotFound(notFindAction: => Tab): TabSelect = {
    this.notFoundAction = () => notFindAction
    this
  }

  def whenNotFound(loadData: => Unit, notFindAction: => Tab): TabSelect = {
    this.notFoundAction = () => notFindAction
    this.loadData = () => loadData
    this
  }

  def select(title: String): Unit =
    root.getTabs.asScala.to(LazyList)
      .find(_.getText == title) match {
      case Some(tab) =>
        root.getSelectionModel.select(tab)
      case None =>
        MainController().insertTab(notFoundAction())
        if (loadData != null)
          loadData()
    }

}